python/lsst/ts/wep/donutSizeCorrelator.py

```diff
@@ -0,0 +1,348 @@
+        resolution : int, optional
+            Resolution of donut diameter in pixels. Default is 4.
+        d_min : int, optional
+            Minimum donut diameter in pixels. Default is 20.
+        d_max : int, optional
+            Maximum donut diameter in pixels. Default is 200.
+        length : int or None, optional
+            Size of stamp to cutout from center of image.
+            Default is None.
+
+        Returns
+        -------
+        np.ndarray
+            Array of diameters in pixels
+        np.ndarray
+            Array of correlation values
+        """
+        # Crop and bin the image
+        image = self.cropAndBinImage(
+            image=image,
+            length=length,
+            pad=min(d_max // 2, np.min(image.shape) // 4),
+            binning=resolution,
+        )
+
+        # Start with minimum diameter
+        diam = d_min
+
+        diameters = []
+        correlations = []
+        for diam in np.arange(d_min, d_max + resolution, resolution):
+            # Don't go past d_max
+            if diam > d_max:
+                break
+
+            # Create new template
+            template = self.createDonutTemplate(diam / resolution)
+
+            # Normalize the template
+            template /= template.sum()
+
+            # Calculate max correlation in image
+            corr_img = correlate(image, template)
+            corr = np.nanmax(corr_img)
+
+            # Save values
+            diameters.append(diam)
+            correlations.append(corr)
+
+        diameters = np.array(diameters)
+        correlations = np.array(correlations)
+
+        return diameters, correlations
+
+    def getDonutDiameter(
+        self,
+        image: np.ndarray,
+        d_min: int | None = 10,
+        d_max: int | None = 420,
+    ) -> tuple[int | float, np.ndarray, np.ndarray]:
+        """Get donut diameter by correlating the image.
+
+        Parameters
+        ----------
+        image : np.ndarray
+            Image containing donuts
+        d_min : int or None, optional
+            Minimum donut diameter in pixels. Default is 10.
+        d_max : int or None, optional
+            Maximum donut diameter in pixels. Default is 420.
+
+        Returns
+        -------
+        int or float
+            Estimate of the donut diameter in pixels.
+            If the algorithm fails, it returns a NaN.
+        np.ndarray
+            Array of tested diameters
+        np.ndarray
+            Array of resulting correlations
+        """
+        # Instantiate empty arrays
+        diameters = np.array([], dtype=int)
+        correlations = np.array([])
+
+        # We will progress by powers of 2 in resolution
+        # and decrease resolution when resolution = 5% of diameter
+        # and max diameter we will test is longest side of image
+        L_max = max(image.shape)
+        niter = np.ceil(np.log2(0.05 * L_max)).astype(int)
+
+        for n in range(0, niter + 1):
```

Comment:
so help me understand this, why do we have a loop here and also inside the correlateImage function?

Reply:
Here we iterate over  donut diameter  grids  of different resolution. Then for each grid, correlateImage loops templates corresponding to probed diameters.  `correlateImage` stores the diameters and corresponding correlation values.  That's why here we  append these results of different grids : ```     # Append to our existing arrays
            diameters = np.concatenate((diameters, diam[1:]))
            correlations = np.concatenate((correlations, corr[1:]))```